atproto/identity/identity.go

```diff
@@ -0,0 +1,187 @@
+package identity
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/bluesky-social/indigo/atproto/crypto"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+
+	"github.com/mr-tron/base58"
+)
+
+// API for doing account lookups by DID or handle, with bi-directional verification handled automatically. Almost all atproto services and clients should use an implementation of this interface instead of resolving handles or DIDs separately
+//
+// Handles which fail to resolve, or don't match DID alsoKnownAs, are an error. DIDs which resolve but the handle does not resolve back to the DID return an Identity where the Handle is the special `handle.invalid` value.
+//
+// Some example implementations of this interface could be:
+//   - basic direct resolution on every call
+//   - local in-memory caching layer to reduce network hits
+//   - API client, which just makes requests to PDS (or other remote service)
+//   - client for shared network cache (eg, Redis)
+type Catalog interface {
+	LookupHandle(ctx context.Context, h syntax.Handle) (*Identity, error)
+	LookupDID(ctx context.Context, d syntax.DID) (*Identity, error)
+	Lookup(ctx context.Context, i syntax.AtIdentifier) (*Identity, error)
+	// TODO: add "flush" methods to purge caches?
+}
+
+// Indicates that resolution process completed successfully, but handle does not exist.
+var ErrHandleNotFound = errors.New("handle not found")
+
+// Indicates that handle and DID resolved, but handle points to a DID with a different handle. This is only returned when looking up a handle, not when looking up a DID.
+var ErrHandleNotValid = errors.New("handle resolves to DID with different handle")
+
+// Indicates that resolution process completed successfully, but the DID does not exist.
+var ErrDIDNotFound = errors.New("DID not found")
+
+var DefaultPLCURL = "https://plc.directory"
+
+// Returns a reasonable default Catalog implementation for most use cases
+func DefaultCatalog() Catalog {
+	naive := NewBasicCatalog(DefaultPLCURL)
+	cached := NewCacheCatalog(&naive)
+	return &cached
+}
+
+// Represents an atproto identity. Could be a regular user account, or a service account (eg, feed generator)
+type Identity struct {
+	DID syntax.DID
+
+	// Handle/DID mapping must be bi-directionally verified. If that fails, the Handle should be the special 'handle.invalid' value
+	// TODO: should we make this nullable, instead of 'handle.invalid'?
+	Handle syntax.Handle
```

Comment:
Not all Identities have Handles i.e. lots of FeedGenerators have DID docs at their `https://{feed_host}/.well-known/did.json` and no AKA's defined in them. Do we only want this package to support Identities with handles?

Reply:
This package should support those just fine; they simply have an "invalid" handle. If you never try to use the handle it isn't a problem.

As the comment says, we could have this field be nullable, which would be a bit more efficient (in memory) for cases with no handle. But it is kind of weird to have it be both nullable *and* there exist the special invalid handle. FWIW, I was against the special invalid handle existing and thought that we should have updated all the lexicons to have nullable handles instead, but we didn't want to break things, and having a specific value fails more gracefully in some situations. I could go either way on this: if there is a performance concern i'm fine having it be nullable.

A tangential corner-case with service accounts vs. regular PDS accounts is that they are not mutually exclusive: you can have a did:web feed generator which also has a repo (PDS location and public signing key).

Reply:
we discussed and decided current API is fine